handler/file/range_reader.go

```diff
@@ -0,0 +1,64 @@
+package file
+
+import (
+	"errors"
+	"io"
+)
+
+// rangeReader reads data from one individually retrievable file range.
+type rangeReader struct {
+	// offset is the absolute offset within file where the next read will get
+	// data from.
+	offset    int64
+	reader    io.ReadCloser
+	remaining int64
+}
+
+func (rr *rangeReader) inRange(pos int64) bool {
+	return pos >= rr.offset && pos < rr.offset+rr.remaining
+}
+
+func (rr *rangeReader) writeToN(w io.Writer, readLen int64) (int64, error) {
+	var read int64
+	for readLen > 0 {
+		if rr.remaining == 0 {
+			return read, io.EOF
+		}
+		var n int64
+		var err error
+
+		if readLen >= rr.remaining {
+			// Copy all remaining bytes.
+			n, err = io.Copy(w, rr.reader)
+		} else {
+			// Copy requested number of bytes.
+			n, err = io.CopyN(w, rr.reader, readLen)
+		}
+		if err != nil && !errors.Is(err, io.EOF) {
+			return 0, err
```

Comment:
should we capture error state here and return the same if called again?

Reply:
I wanted to leave that up to the caller to deal with. For example, if the error is something temporary like "device busy", and maybe the caller wants to implement retry logic.